test/unit/tuple_bigref.c

```diff
@@ -5,14 +5,93 @@
 #include <stdio.h>
 
 enum {
-	BIGREF_DIFF = 10,
-	BIGREF_COUNT = 70003,
-	BIGREF_CAPACITY = 107,
+	FEW_REFS = 10,
+	MANY_REFS = 1000,
+	TEST_MAX_TUPLE_COUNT = 1024,
+	RAND_TEST_ROUNDS = 1024 * 1024,
 };
 
 static char tuple_buf[64];
 static char *tuple_end = tuple_buf;
 
+size_t tuple_count = 0;
+size_t tuple_delete_error_count = 0;
+static struct tuple *allocated_tuples[TEST_MAX_TUPLE_COUNT];
+
+static struct tuple_format *patched_format;
+static struct tuple * (*save_tuple_new)(struct tuple_format *format,
+					const char *data, const char *end);
+static void (*save_tuple_delete)(struct tuple_format *format,
+				 struct tuple *tuple);
+
+static void patch_format();
+static void restore_format();
+
+/**
+ * Check an expression that has human readable explanation.
+ * If the expression is false - print the explanation and terminate execution.
+ *
+ * This function is useful for invariant checks that, once failed, makes
+ * senseless further work.
+ *
+ * @param expression
+ * @param explanation
+ */
+static void
+ensure(bool expression, const char *explanation)
```

Comment:
We already have fail_if/fail_unless helpers doing the same thing. Please use them instead.

Reply:
But fail_if/fail_unless don't have an error message. For example in this test `fail_if` will output something like:
```Test failed: pos < tuple_count is true at file.. line.. func.. ```
while ensure will give 
```Fatal check failed: The tuple begin deleted *was* allocated"```
Are you sure that fail_if gives more informative message?

Reply:
s/begin/being

It looks more like an assertion to me, not a test case check. If you tried to free a tuple that was not allocated, you would have likely already failed anyway trying to dereference uninitialized/freed memory. Let's use fail_if and add a comment instead. I don't like introducing another assert-like helper.

Reply:
ok.